lib/safira/roulette/prize.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Safira.Roulette.Prize do
+  use Ecto.Schema
+  use Arc.Ecto.Schema
+  import Ecto.Changeset
+
+  alias Safira.Accounts.Attendee
+
+  schema "prizes" do
+    field :name, :string
+    field :stock, :integer
+    field :max_amount_per_attendee, :integer
+    field :probability, :float
+    field :avatar, Safira.Avatar.Type
+
+    many_to_many :attendees, Attendee, join_through: AttendeePrize
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(prize, attrs) do
+    prize
+    |> cast(attrs, [:name, :stock, :max_amount_per_attendee, :probability])
+    |> cast_attachments(attrs, [:avatar])
+    |> validate_required([:name, :stock, :max_amount_per_attendee, :probability])
+    |> (&validate_number(&1, :stock,
+          greater_than_or_equal_to: fetch_field(&1, :max_amount_per_attendee) |> elem(1)
+        )).()
```

Comment:
Could you explain to me the final part of the `validate_number` function? why is `greater_than_or_equal_to: fetch_field(&1, :max_amount_per_attendee` being fed to `elem(1)`, how does the first condition yield a tuple? Also, what does the `.()` mean?

Reply:
Just to put into context, in this changeset verification it was intended to check that the value of the `:stock` had to be greater than or equal to the `:max_amount_per_attendee`. To do that we had to access the value of `:max_amount_per_attendee` which is in the changeset structure. The problem is that you can't simply access the changeset structure because its value comes from the pipe.

Dividing the answer:

1. You can pipe into an anonymous function but you have to express it as a call to that function. In elixir `fn(x) -> 1 + x end` is identical to `&(1 + &1)`. But if you name this function as x then basically what you'd be doing is `|> x.()` hence the `.()`. If you remove the .`()`, it's the same as having the function and not calling it.

2. [fetch_field/2](https://hexdocs.pm/ecto/Ecto.Changeset.html#fetch_field/2) is a function that fetches the given field from changes or from the data and returns a tuple. The `elem/2` accesses this tuple by index and since I want to access the second index, I simply call `elem(1)`.

Reply:
okay, I see, my problem was not knowing that `fetch_field` returns a tuple, I tought it returned only the desired value, now it makes sense

Reply:
Thinking of it: 
Assuming that the `:stock` field will be decremented every time an attendee wins a prize won't there be a time where the available stock will be lower than the `max_amount_per_attendee`, which will make it impossible to win the last few pieces of any prize? For example, let's assume one of the prizes is a pen, with an initial stock off 500 and a `max_amount_per_attendee` of 10. Now, later, 490 of those pens have now already been won (which means we only have a stock of 10) and another one is won by an user, won't the changeset change the `:stock` field to 9, which will then be claimed as invalid by the `greater_than_or_equal_to:` validation since 9 <10 max_amount_per_attendee?

Or will this changeset only ever by used by the initial creation of a prize and another different changeset will be used when a prize is actually won?

Reply:
yes, we'll need a `update_changeset` for that case 👍